src/inc/til/spsc.h

```diff
@@ -0,0 +1,526 @@
+                if (end > beg)
+                {
+                    std::destroy(_data + beg, _data + end);
+                }
+                else if (differentRevolution)
+                {
+                    std::destroy(_data, _data + end);
+                    std::destroy(_data + beg, _data + _capacity);
+                }
+
+                free_raw_memory(_data);
+            }
+
+            void drop_producer()
+            {
+                drop(_producer);
+            }
+
+            void drop_consumer()
+            {
+                drop(_consumer);
+            }
+
+            acquisition producer_acquire(size_type slots) noexcept
+            {
+                return acquire(_producer, _consumer, revolution_flag, slots);
+            }
+
+            void producer_release(acquisition acquisition) noexcept
+            {
+                release(_producer, acquisition);
+            }
+
+            acquisition consumer_acquire(size_type slots) noexcept
+            {
+                return acquire(_consumer, _producer, 0, slots);
+            }
+
+            void consumer_release(acquisition acquisition) noexcept
+            {
+                release(_consumer, acquisition);
+            }
+
+            T* data() const noexcept
+            {
+                return _data;
+            }
+
+        private:
+            void drop(atomic_size_type& mine)
+            {
+                const auto myPos = mine.load(std::memory_order_relaxed);
+                mine.store(myPos | drop_flag, std::memory_order_release);
+                mine.notify_one();
+
+                // The first time SPSCBase is dropped (destroyed) we'll set
+                // the flag to true and get false, causing us to return early.
+                // Only the second time we'll get true.
+                // --> The contents are only deleted when both sides have been dropped.
+                if (_eitherSideDropped.exchange(true))
+                {
+                    delete this;
+                }
+            }
+
+            // NOTE: waitMask MUST be either 0 (consumer) or revolution_flag (producer).
+            acquisition acquire(atomic_size_type& mine, atomic_size_type& theirs, size_type waitMask, size_type slots) noexcept
+            {
+                // We can always safely load "mine" as we're the only one writing to it.
+                size_type myPos = mine.load(std::memory_order_relaxed);
+                size_type theirPos;
+
+                while (true)
+                {
+                    // On the other hand we need to safely & atomically fetch the foreign "theirs" value.
+                    theirPos = theirs.load(std::memory_order_acquire);
+                    if ((myPos ^ theirPos) != waitMask)
+                    {
+                        break;
+                    }
+
+                    theirs.wait(theirPos, std::memory_order_acquire);

```

Comment:
`memory_order_acquire` only observes visible side effects of another thread if it wrote data using `memory_order_release`.
Basically "acquire" loads observe prior "release" writes on the same atomic. Any other kind of memory order (except for `seq_cst`) will not interfere with this at all. For instance if you concurrently write using the "release" order, the writes will not synchronize with each other and the threads might still read stale data from each other. Only if you read the data back using the "acquire" order you'll establish a happened-before order and observe all visible side effects of all other writers at the time of their last write.

Reply:
But in the loop, after the `theirs.wait` returns, we're always going to do a second `theirs.load(std::memory_order_acquire)` anyway, so couldn't this one be relaxed? Because if we do relaxed then acquire, and then do the read, it should be the same thing as if we do acquire then acquire, and then do the read?

Reply:
To extend on your comment's edit: The new `std::atomic<T>::wait(old, order)` will try to block the thread until the atomic's value is different compared to the `old` argument. If I'd hypothetically call `wait()` with a relaxed order, there's technically no guarantee `wait()` ever returns, because using relaxed order there's technically no guarantee that the blocked thread will ever observe any writes to the atomic. As such `wait()` should AFAICS only be used with acquire/release or sequentially consistent order. (Windows' and Linux' futexes (`WaitOnAdress` for Windows) both only support sequentially consistent ordering anyways btw.)

Reply:
I'm having a hard time trying to read the documentation about this 😂, but I think that I've found some interesting things on the Internet that make me doubt about "using relaxed order there's technically no guarantee that the blocked thread will ever observe any writes to the atomic":

On https://en.cppreference.com/w/cpp/atomic/memory_order:

> Release-Acquire ordering
> If an atomic store in thread A is tagged memory_order_release and an atomic load in thread B from the same variable is tagged memory_order_acquire, all memory writes (non-atomic and relaxed atomic) that happened-**before** the atomic store from the point of view of thread A, become visible side-effects in thread B. That is, once the atomic load is completed, thread B is guaranteed to see everything thread A wrote to memory.

So this seems to be about not seeing the atomic store itself, but seeing things that happened _before_ the atomic store.

This answer on SO https://stackoverflow.com/questions/48124031/stdmemory-order-relaxed-atomicity-with-respect-to-the-same-atomic-variable says:
> The following are 2 typical scenario's whereby ordering constraints on an atomic operation can be omitted (i.e. by using std::memory_order_relaxed):
> 1. Memory ordering is not necessary because there are no dependencies on other operations, or as a commenter puts it, (..) is not part of an invariant involving other memory locations.

This question on SO https://stackoverflow.com/questions/31150809/understanding-memory-order-relaxed (the first part of the question) has an answer which says that even with relaxed, the second thread will read the value written by the first thread.

What I understand is that `std::memory_order`s don't guarantee anything about when the thread will observe the write, they just guarantee that if it does, then _other reads/writes_ around cannot be reordered.

Reply:
No, you’re understanding this correctly. Atomics guarantee that writes propagate to reads within finite time. But a relaxed memory order doesn’t guarantee the immediate synchronization of your value. If I write a value to the atomic and Call `notify_one()` it could theoretically immediately resume the other, waiting thread which will then observe a stale read due to the relaxed ordering.
Relaxed ordering also allows the compiler to reorder instructions it thinks are unrelated.

As such I believe the SO answers you linked technically aren’t correct (they‘re definitely correct for X86 though as it only has one kind of atomics anyways and there’s no difference between relaxed and any other memory ordering).
To support my argument I found this: https://stackoverflow.com/questions/43749985/is-it-possible-that-a-store-with-memory-order-relaxed-never-reaches-other-thread

Reply:
You know what? I've thought about it again and I believe you're correct... It wouldn't make sense if it was allowed for wait/notify calls to be reordered. The spec (P1135R6) doesn't explicitly say so, but a `x.store(val)` followed by `x.notify_one()` both access the same atomic variable `x` and thus can't be reordered within the same thread. (Because that's what I was afraid of before).
I'm changing the wait() to use relaxed ordering then. I could also modify the load above to use relaxed ordering in that case, as long as I have a `std::atomic_thread_fence(acquire)` before I return from `acquire()`. But benchmarking didn't show any meaningful difference, so I'll stick with the current acquire load. (I think the acquire loop won't be spinning tightly anyways, thanks to the `wait()` call.)
Thanks for your comments! It gave me quite a bit to think about this relatively hairy topic.